rcl/test/rcl/test_subscription.cpp

```diff
@@ -178,14 +177,15 @@ TEST_F(CLASSNAME(TestSubscriptionFixture, RMW_IMPLEMENTATION), test_subscription
   {
     test_msgs__msg__BasicTypes msg;
     test_msgs__msg__BasicTypes__init(&msg);
-    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
-    {
-      test_msgs__msg__BasicTypes__fini(&msg);
-    });
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+        test_msgs__msg__BasicTypes__fini(&msg);
+      });
     ret = rcl_take(&subscription, &msg, nullptr, nullptr);
     ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
     ASSERT_EQ(42, msg.int64_value);
   }
+  ret = rcl_subscription_fini(&subscription, this->node_ptr);
```

Comment:
In this case, we should probably use the `OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT` macro because there are some `ASSERT_*` statements between the subscription init and the fini. If an assert fails, then this finalize statement is not reached.

Reply:
Thank you for careful review!
Right, if ASSERT fires then fini isn't be called.
Hmm, I think the problem is the scope of subscription is too wide and re-used.
And OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT may not be suitable in this situation.

If we add this macro after last subscription_init, then fini is called for now even when ASSERT fies.
But in someday if someone add other tests in this function after last fini, leak may be happen.

I think the right solution is either of:
(1) create a new function for test from last init to last fini.
(2) use local scope (as my 1st commit)
(3) don't care...  because if ASSERT fires, there is something more bad

Item (1) looks better for me, how do you think?

Reply:
Options (1) or (2) are okay with me.

Reply:
Sorry for the back-and-forth. Your first commit was perfectly fine, I was just being nit-picky.

Reply:
No problem, I'm enjoying discussion with you! Thank you very much.
Well, for code clearness, I'll do (1), please wait.

Thank you.

Reply:
I introduced new function `test_subscription_nominal_valid`.
Here is the test result.
The new function is called. And the fact that there is no line under "[  PASSED  ]" means to pass asan test.

```
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from TestSubscriptionFixture__rmw_fastrtps_cpp
[ RUN      ] TestSubscriptionFixture__rmw_fastrtps_cpp.test_subscription_nominal
[       OK ] TestSubscriptionFixture__rmw_fastrtps_cpp.test_subscription_nominal (17 ms)
[ RUN      ] TestSubscriptionFixture__rmw_fastrtps_cpp.test_subscription_nominal_valid
[       OK ] TestSubscriptionFixture__rmw_fastrtps_cpp.test_subscription_nominal_valid (1011 ms)
[ RUN      ] TestSubscriptionFixture__rmw_fastrtps_cpp.test_subscription_nominal_string
[       OK ] TestSubscriptionFixture__rmw_fastrtps_cpp.test_subscription_nominal_string (1016 ms)
[----------] 3 tests from TestSubscriptionFixture__rmw_fastrtps_cpp (2044 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (2044 ms total)
[  PASSED  ] 3 tests.
```